### ISSUE TITLE
terraform console crashes when using data sources

### ISSUE THREAD

**Issue by eilgin** [hashicorp/terraform#16445]
### Terraform Version
Run `terraform -v` to show the version. If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.

Used version:
```
$ terraform --version
Terraform v0.10.7
```

### Terraform Configuration Files
```hcl
data "google_compute_lb_ip_ranges" "ranges" {}
```

### Crash Output
https://gist.github.com/eilgin/430f690c6c7269c83da18b8f6bad2f88

### Expected Behavior
I expect the console to fetch and returns the list of IP ranges used by the google load balancer and health checkers. I've also try with aws data sources and the same happens. It may be possible that the `terraform console` wasn't made to check against data sources ?

### Actual Behavior
It immediately crashes.

### Steps to Reproduce
Please list the full steps required to reproduce the issue, for example:
1. `terraform console`
2. type and try to fetch results from a data source, e.g., `data.google_compute_lb_ip_ranges.ranges.network`

### Important Factoids
ø

### References
I didn't find anything relevant with the following filters: `is:issue is:open console data`

**Comment by pdecat** [comment]
It no longer crashes with terraform 0.11.2 when the datasource is not present in the state:

```
# echo data.google_compute_lb_ip_ranges.ranges.network | terraform console
Resource 'data.google_compute_lb_ip_ranges.ranges' not found for variable 'data.google_compute_lb_ip_ranges.ranges.network'
```


Also, to get values, you need to refresh the state:

```
# echo data.google_compute_lb_ip_ranges.ranges.network | terraform console
[
  209.85.152.0/22,
  209.85.204.0/22,
  35.191.0.0/16
]
```

**Comment by AnthonyWC** [comment]
Just came across I think the same issue:
```Terraform v0.11.3```

```echo data.aws_caller_identity.current.account_id | terraform console                            
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x173a944]

goroutine 1 [running]:
github.com/hashicorp/terraform/config/module.(*Tree).Config(...)
	/private/tmp/terraform-20180131-15725-59s40c/terraform-0.11.3/src/github.com/hashicorp/terraform/config/module/tree.go:78
github.com/hashicorp/terraform/terraform.(*Interpolater).resourceVariableInfo(0xc4202f1980, 0xc4201dd890, 0xc4203c0820, 0x9, 0x34573c0, 0x2af6070, 0x2afd0f7)
	/private/tmp/terraform-20180131-15725-59s40c/terraform-0.11.3/src/github.com/hashicorp/terraform/terraform/interpolate.go:756 +0x44
github.com/hashicorp/terraform/terraform.(*Interpolater).computeResourceVariable(0xc4202f1980, 0xc4201dd890, 0xc4203c0820, 0x0, 0x0, 0x0)
	/private/tmp/terraform-20180131-15725-59s40c/terraform-0.11.3/src/github.com/hashicorp/terraform/terraform/interpolate.go:459 +0x14e
github.com/hashicorp/terraform/terraform.(*Interpolater).valueResourceVar(0xc4202f1980, 0xc4201dd890, 0xc4202e5382, 0x2b, 0xc4203c0820, 0xc4203e17d0, 0xc4203e1740, 0xc4203e1740)
	/private/tmp/terraform-20180131-15725-59s40c/terraform-0.11.3/src/github.com/hashicorp/terraform/terraform/interpolate.go:252 +0x68
github.com/hashicorp/terraform/terraform.(*Interpolater).Values(0xc4202f1980, 0xc4201dd890, 0xc4203e17a0, 0x0, 0xc4203f8508, 0xc4202e5380)
	/private/tmp/terraform-20180131-15725-59s40c/terraform-0.11.3/src/github.com/hashicorp/terraform/terraform/interpolate.go:86 +0x82f
github.com/hashicorp/terraform/repl.(*Session).handleEval(0xc4201dda98, 0xc4202e5350, 0x2b, 0x2b, 0x37179c8, 0xc4202e5350, 0x2b)
	/private/tmp/terraform-20180131-15725-59s40c/terraform-0.11.3/src/github.com/hashicorp/terraform/repl/session.go:55 +0x2b7
github.com/hashicorp/terraform/repl.(*Session).Handle(0xc4201dda98, 0xc4202e5350, 0x2b, 0x2b, 0xc4202e5350, 0x2b, 0x3681a1b)
	/private/tmp/terraform-20180131-15725-59s40c/terraform-0.11.3/src/github.com/hashicorp/terraform/repl/session.go:35 +0xa9
github.com/hashicorp/terraform/command.(*ConsoleCommand).modePiped(0xc4203f6d00, 0xc4201dda98, 0x5107f60, 0xc4203e1650, 0xc4203e13b0)
	/private/tmp/terraform-20180131-15725-59s40c/terraform-0.11.3/src/github.com/hashicorp/terraform/command/console.go:108 +0x115
github.com/hashicorp/terraform/command.(*ConsoleCommand).Run(0xc4203f6d00, 0x0, 0x0, 0x0, 0xc4202e7480)
	/private/tmp/terraform-20180131-15725-59s40c/terraform-0.11.3/src/github.com/hashicorp/terraform/command/console.go:97 +0x54a
github.com/hashicorp/terraform/vendor/github.com/mitchellh/cli.(*CLI).Run(0xc42034bb80, 0xc42034bb80, 0xc4201ddd18, 0x1)
	/private/tmp/terraform-20180131-15725-59s40c/terraform-0.11.3/src/github.com/hashicorp/terraform/vendor/github.com/mitchellh/cli/cli.go:255 +0x1eb
main.wrappedMain(0x0)
	/private/tmp/terraform-20180131-15725-59s40c/terraform-0.11.3/src/github.com/hashicorp/terraform/main.go:214 +0xbda
main.realMain(0x0)
	/private/tmp/terraform-20180131-15725-59s40c/terraform-0.11.3/src/github.com/hashicorp/terraform/main.go:99 +0xbb
main.main()
	/private/tmp/terraform-20180131-15725-59s40c/terraform-0.11.3/src/github.com/hashicorp/terraform/main.go:35 +0x38



!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!```

**Comment by apparentlymart** [comment]
Hi all! Sorry for this broken behavior, and sorry for the slow response.

The expression evaluation subsystem has been completely overhauled in master in preparation for the forthcoming v0.12.0 release. I've just verified this is fixed in v0.12.0-alpha1 using the following configuration:

```hcl
provider "aws" {
  region = "us-west-2"
}

data "aws_caller_identity" "current" {
}
```

```
$ terraform apply
data.aws_caller_identity.current: Refreshing state...

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
```

```
$ terraform console
> data.aws_caller_identity.current
{
  "account_id" = "123456"
  "arn" = "arn:aws:iam::123456:user/apparentlymart"
  "id" = "2018-10-30 00:10:09.771694903 +0000 UTC"
  "user_id" = "AIDABLAHBLAH"
}
> data.aws_caller_identity.current.account_id
123456
```

As @pdecat noted above, it's important to run a command that allows Terraform to read values from the data source first. Failing to do this now returns a better error message, however:

```
$ terraform console
> data.aws_caller_identity.current

Error: Result depends on values that cannot be determined until after "terraform apply".

> 
```

Since this is fixed in the master branch, I'm going to close this issue out now. The fix will be included in the forthcoming v0.12.0 final release. Thanks for reporting this, and sorry again for the delay in responding.
